query/raw_result.go

```diff
@@ -131,121 +74,94 @@ func (r *RawResults) IsComplete() bool { return r.LimitReached() && !r.IsOrdered
 
 // MinOrderByValues returns the orderBy field values for the smallest result in
 // the result collection.
-func (r *RawResults) MinOrderByValues() []field.ValueUnion {
+// NB: Determining the min result requires a linear scan and more expensive than determining
+// the max result values. However this is okay because the limit is usually not too large so
+// linear scan is tolerable. Additionally from an algorithm perspective, keeping the results
+// in a heap and sorting at the end is strictly better than always keeping the results sorted
+// (2NlogN + N vs 3NlogN + 2N per segment), so even with the additional O(N) complexity of
+// a linear scan this is still cheaper than always keeping results sorted, and the resulting
+// data structure is easier to maintain.
+func (r *RawResults) MinOrderByValues() field.Values {
 	if r.Len() == 0 {
 		return nil
 	}
-	return r.Data[0].OrderByValues
+	if r.minOrderByValuesReady {
+		return r.minOrderByValues
+	}
+	data := r.Ordered.RawData()
+	minOrderByValues := data[0].OrderByValues
+	for i := 1; i < len(data); i++ {
+		if r.ValuesLessThanFn(data[i].OrderByValues, minOrderByValues) {
+			minOrderByValues = data[i].OrderByValues
+		}
+	}
+	r.minOrderByValues = minOrderByValues
+	r.minOrderByValuesReady = true
+	return minOrderByValues
 }
 
 // MaxOrderByValues returns the orderBy field values for the largest result in
 // the result collection.
-func (r *RawResults) MaxOrderByValues() []field.ValueUnion {
+func (r *RawResults) MaxOrderByValues() field.Values {
 	if r.Len() == 0 {
 		return nil
 	}
-	return r.Data[r.Len()-1].OrderByValues
+	// NB: The "min" element in `Ordered` is actually the largest element among the top N results
+	// because the heap uses `ResultRverseLessThanFn` for ordering purposes.
```

Comment:
ResultR**e**verseLessThanFn

Reply:
Also, this whole thing is kinda confusing (min being the max element), maybe rename `Min()` to `Top()`? `Min()` feels misleading while `Top()` can semantically mean anything.

Reply:
Sounds good can do.